pkg/splunk/resources/util.go

```diff
@@ -197,8 +198,86 @@ func CompareEnvs(a []corev1.EnvVar, b []corev1.EnvVar) bool {
 	return false
 }
 
-// SortVolumeMounts returns a sorted list of Kubernetes VolumeMounts.
-func SortVolumeMounts(mounts []corev1.VolumeMount) []corev1.VolumeMount {
+// sortVolumes returns a sorted list of Kubernetes Volumes.
+func sortVolumes(volumes []corev1.Volume) []corev1.Volume {
+	sorted := make([]corev1.Volume, len(volumes))
+	copy(sorted, volumes)
+	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Name < sorted[j].Name })
+	return sorted
+}
+
+func compareSecretVolume(a, b *corev1.SecretVolumeSource) bool {
+	if a.SecretName != b.SecretName {
+		return true
+	}
+
+	if a.DefaultMode != nil && b.DefaultMode != nil {
+		if *a.DefaultMode != *b.DefaultMode {
+			return true
+		}
+	}
+	return false
+}
+
+func compareConfigMapVolume(a, b *corev1.ConfigMapVolumeSource) bool {
+	if a.Name != b.Name {
+		return true
+	}
+
+	if a.DefaultMode != nil && b.DefaultMode != nil {
+		if *a.DefaultMode != *b.DefaultMode {
+			return true
+		}
+	}
+
+	if a.LocalObjectReference != b.LocalObjectReference {
+		return true
+	}
+
+	if a.Optional != nil && b.Optional != nil {
+		if *a.Optional != *b.Optional {
+			return true
+		}
+	}
+	return false
+}
+
+// CompareVolumes is a generic comparer of two Kubernetes Volumes.
+// It returns true if there are material differences between them, or false otherwise.
+func CompareVolumes(a []corev1.Volume, b []corev1.Volume) bool {
+	// first, check for short-circuit opportunity
+	if len(a) != len(b) {
+		return true
+	}
+
+	// make sorted copies of a and b
+	aSorted := sortVolumes(a)
+	bSorted := sortVolumes(b)
+
+	// iterate elements, checking for differences
+	for n := range aSorted {
+
+		// Compare secret volume separately because defaultMode is optional
+		if aSorted[n].VolumeSource.Secret != nil {
```

Comment:
Does reflect.DeepEqual() not include these in the comparisons? Not clear to me why just calling that isn't sufficient.

Reply:
The issue is the optional field "DefaultMode" . k8s initialize this field if nil (https://github.com/kubernetes/kubernetes/blob/94837daab20ce5b3fee00f26483a4d64c4d7b9e4/pkg/apis/core/v1/defaults.go#L237). So every time we compare, it will not be equal. Another option is we initialize this field ourselves when we create the volumes internally AND when the user specifies it via CR. I have another PR for this but I think this might be cleaner since we are doing it in 1 place here.

Reply:
Another thing we need to consider is we may want to move to having volumeClaimTemplates similar to serviceTemplate, which would allow for more flexibility as contemplated in #50. This would break the assumptions you're making in here.

My preference would be to initialize these fields and keep the comparison methods strict. Wonder if would make sense to call these SetDefaults methods ourselves to ensure they stay in sync?

Reply:
That was my thought initially to keep the comparison strict. Let me bring up the PR for this case.